elki/src/main/java/de/lmu/ifi/dbs/elki/math/statistics/dependence/MCDEDependenceMeasure.java

```diff
@@ -0,0 +1,173 @@
+package de.lmu.ifi.dbs.elki.math.statistics.dependence;
+
+
+import de.lmu.ifi.dbs.elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import de.lmu.ifi.dbs.elki.utilities.documentation.Reference;
+import de.lmu.ifi.dbs.elki.utilities.exceptions.AbortException;
+import de.lmu.ifi.dbs.elki.utilities.random.RandomFactory;
+import de.lmu.ifi.dbs.elki.utils.containers.RankStruct;
+import java.util.Arrays;
+import java.util.Random;
+
+
+/**
+ * Implementation of bivariate Monte Carlo Density Estimation as described in
+ * Edouard Fouché & Klemens Böhm<br>
+ * Monte Carlo Density Estimation<br>
+ * Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)
+ *
+ * This is an abstract class. In order to use MCDE extend it and implement an appropriate statistical test that
+ * returns a p-value and index structure for efficient computation of the statistical test.
+ *
+ * The instantiation of MCDE based on the Mann-Whitney U test is called MWP (as described in the paper).
+ * See McdeMwpDependenceMeasure.
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+        title = "Monte Carlo Density Estimation", //
+        booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)",
+        url = "https://doi.org/10.1145/3335783.3335795", //
+        bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+
+public abstract class MCDEDependenceMeasure<R extends RankStruct> extends AbstractDependenceMeasure {
```

Comment:
I'm wondering if the abstract class is useful as is; because there is only one subclass, right?
This would allow eliminating the entire `RankStruct` construct? For long-term maintainability, I have been repeatedly trying to simplify class hierarchies, removing generics, etc.

Reply:
Yes I adjusted it. We now have the abstract `MCDETest` and its subclass `MWPTest` that creates the index and computes the statistical test. They get passed to `MCDEDependenceMeasure` They also contain `RankStruct` and `MwpIndex` as inner classes. The reason I used inheritance is that the way we compute MWP is very tailored to the whole MCDE framework to be efficient therefore the tests are not really usefull outside of it. `McdeMwpDependenceMeasure` has been deleted.